Issue when running plugin in a multi-project build

Hello,

I'm using the plugin in 2 separate subprojects of a root project. These are: _libraries/uicomponents_ and _features/userdetails_. It's a JavaFX project and I'm adding a functionality to define different styles per screen. Each screen has its own subproject.

Right now, the plugin configuration in _build.gradle_ is the same in both subprojects:


```
jar {
    dependsOn compileSass
    from compileSass.outputDir
}

sass {
    noAutoCopy()
}

compileSass {
    sourceDir = project.file("${projectDir}/src/main/resources")
    outputDir = project.file("${buildDir}/resources/main")
    loadPath project.file("${rootDir}/common/sass-lib")
    sourceMap = none
}
```
However, when building the whole project there goes an error:

```
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :jar
> Task :assemble
> Task :compileTestJava NO-SOURCE
> Task :processTestResources NO-SOURCE
> Task :testClasses UP-TO-DATE
> Task :test NO-SOURCE
> Task :check UP-TO-DATE
> Task :build
> Task :libraries:core:compileJava
> Task :libraries:core:processResources NO-SOURCE
> Task :libraries:core:classes
> Task :libraries:core:jar
> Task :libraries:uicomponents:compileJava
> Task :libraries:uicomponents:processResources NO-SOURCE
> Task :libraries:uicomponents:classes
> Task :libraries:uicomponents:downloadSass UP-TO-DATE
> Task :libraries:uicomponents:installSass UP-TO-DATE
> Task :features:home:processResources
> Task :features:settings:processResources
> Task :features:userdetails:processResources
> Task :features:userdetails:downloadSass UP-TO-DATE

> Task :libraries:uicomponents:compileSass FAILED
The process cannot access the file because it is being used by another process.

> Task :features:userdetails:installSass
Execution optimizations have been disabled for task ':features:userdetails:installSass' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'C:\projects\javafx-app-template\.gradle\sass\archive\dart-sass-1.54.0-windows-x64.zip'. Reason: Task ':features:userdetails:installSass' uses this output of task ':libraries:uicomponents:downloadSass' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.
14 actionable tasks: 9 executed, 5 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libraries:uicomponents:compileSass'.
> A failure occurred while executing io.miret.etienne.gradle.sass.CompileSassWorkAction
   > Process 'command 'C:\projects\javafx-app-template\.gradle\sass\1.54.0\dart-sass\sass.bat'' finished with non-zero exit value 1
```

Everything is ok when running subprojects build separately.

I found a workaround for the problem and this is setting separate directory for the sass exec files per project, like this:
```
sass {
    directory = file("${rootDir}/.gradle/sass/${name}")
    noAutoCopy()
}
```
but when having a lot of subprojects I can imagine this being space consuming.


I'm using Gradle 7.3.3 version.

I would really appreciate some help on the matter :)

--- a/README.md
+++ b/README.md
@@ -115,7 +115,10 @@ plugin and its `jar` task.
 
 ### In a multi-projects build
 
-You can apply this plugin to several subprojects of a multi-projects build.
+You can apply this plugin to a multi-projects build.
+If only one subproject need it, just apply it to said subproject.
+Otherwise, apply it to the root project,
+or to a common ancestor of all subprojects that need it.
 See [samples/multi-projects](samples/multi-projects/build.gradle).
 
 ### Watching for changes

--- a/samples/multi-projects/build.gradle
+++ b/samples/multi-projects/build.gradle
@@ -1,3 +1,15 @@
+/*
+ * In case several subprojects need this plugin, it is recommended to apply and
+ * configure it on the root project. For one, it allows you to share
+ * the sass version (among other configuration). For another, it will save
+ * disk space as dart-sass will be installed once per project where this plugin
+ * is applied.
+ */
 plugins {
-  id 'io.miret.etienne.sass' version '1.4.2' apply false
+  id 'io.miret.etienne.sass' version '1.4.2'
+}
+
+sass {
+  version = "1.54.4"
+  // Your customization here
 }

--- a/samples/multi-projects/subproject-a/build.gradle
+++ b/samples/multi-projects/subproject-a/build.gradle
@@ -1,4 +1,8 @@
+/* Don't reapply this plugin on subprojects if already done in root project. */
 plugins {
   id 'war'
-  id 'io.miret.etienne.sass'
+}
+
+compileSass {
+  // Configuration for this subproject "compileSass" task.
 }

--- a/samples/multi-projects/subproject-b/build.gradle
+++ b/samples/multi-projects/subproject-b/build.gradle
@@ -1,4 +1,8 @@
+/* Don't reapply this plugin on subprojects if already done in root project. */
 plugins {
   id 'war'
-  id 'io.miret.etienne.sass'
+}
+
+compileSass {
+  // Configuration for this subproject "compileSass" task.
 }

--- a/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/app/build.gradle
+++ b/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/app/build.gradle
@@ -1,8 +1,0 @@
-plugins {
-  id 'io.miret.etienne.sass'
-}
-
-sass {
-  version = '42.0'
-  baseUrl = System.getenv("URL")
-}

--- a/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/build.gradle
+++ b/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/build.gradle
@@ -1,0 +1,8 @@
+plugins {
+  id 'io.miret.etienne.sass'
+}
+
+sass {
+  version = '42.0'
+  baseUrl = System.getenv("URL")
+}

--- a/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/lib/build.gradle
+++ b/src/functionalTest/resources/io/miret/etienne/gradle/sass/multi-project/lib/build.gradle
@@ -1,8 +1,0 @@
-plugins {
-  id 'io.miret.etienne.sass'
-}
-
-sass {
-  version = '42.0'
-  baseUrl = System.getenv("URL")
-}

--- a/src/main/java/io/miret/etienne/gradle/sass/CompileSass.java
+++ b/src/main/java/io/miret/etienne/gradle/sass/CompileSass.java
@@ -4,6 +4,7 @@ import lombok.Getter;
 import lombok.Setter;
 import org.apache.tools.ant.taskdefs.condition.Os;
 import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.tasks.*;
 import org.gradle.workers.WorkQueue;
@@ -86,16 +87,29 @@ public class CompileSass extends DefaultTask {
   @InputFile
   public File getExecutable () {
     String command = Os.isFamily (Os.FAMILY_WINDOWS) ? "sass.bat" : "sass";
-    SassGradlePluginExtension sassExtension = getProject ()
-        .getExtensions ()
-        .findByType (SassGradlePluginExtension.class);
-    assert sassExtension != null;
+    SassGradlePluginExtension sassExtension = findExtension();
     return sassExtension.getDirectory ()
         .toPath ()
         .resolve (sassExtension.getVersion ())
         .resolve ("dart-sass")
         .resolve (command)
         .toFile ();
+  }
+
+  private SassGradlePluginExtension findExtension() {
+    Project project = getProject();
+    SassGradlePluginExtension extension = null;
+    while (extension == null && project != null) {
+      extension = project.getExtensions()
+          .findByType(SassGradlePluginExtension.class);
+      project = project.getParent();
+    }
+    if (extension == null) {
+      throw new IllegalStateException(
+          "SassGradlePluginExtension wasn't registered in any parent project."
+      );
+    }
+    return extension;
   }
 
   public void loadPath (File loadPath) {

--- a/src/main/java/io/miret/etienne/gradle/sass/SassGradlePlugin.java
+++ b/src/main/java/io/miret/etienne/gradle/sass/SassGradlePlugin.java
@@ -43,6 +43,14 @@ public class SassGradlePlugin implements Plugin<Project> {
           task.from (downloadedFiles);
           task.into (new File (extension.getDirectory (), extension.getVersion ()));
         });
+    compileSass(project, extension, installSass);
+  }
+
+  private void compileSass(
+      Project project,
+      SassGradlePluginExtension extension,
+      TaskProvider<?> installSass
+  ) {
     TaskProvider<CompileSass> compileSass = project.getTasks ()
         .register ("compileSass", CompileSass.class, task -> {
           task.setDescription ("Compile sass and scss.");
@@ -56,6 +64,9 @@ public class SassGradlePlugin implements Plugin<Project> {
             task.from (compileSass.map (CompileSass::getOutputDir));
           }
         });
+    project.getSubprojects().forEach(
+        subProject -> compileSass(subProject, extension, installSass)
+    );
   }
 
   private String archiveName (String version) {

